### PR TITLE
Fiqare perseo-fe improvements

### PR DESCRIPTION
**Pull request by fiqare-emergya-dev**
This pull request contains the improvements made by the Emergya team for perseo-fe. These improvements are part of the fiQare project, which is based on ISO 25010 to improve software quality. More info: https://fiqare.eu/

There are 72 issues:
Rule | Severity | File
javascript:S2611 | CRITICAL | fiware-perseo:config.js
javascript:S2611 | CRITICAL | fiware-perseo:config.js
javascript:S2611 | CRITICAL | fiware-perseo:config.js
javascript:S2611 | CRITICAL | fiware-perseo:config.js
javascript:S1226 | MINOR | fiware-perseo:lib/alarm.js
javascript:S1226 | MINOR | fiware-perseo:lib/alarm.js
javascript:S1226 | MINOR | fiware-perseo:lib/alarm.js
javascript:S1226 | MINOR | fiware-perseo:lib/alarm.js
javascript:S1226 | MINOR | fiware-perseo:lib/alarm.js
javascript:S1226 | MINOR | fiware-perseo:lib/alarm.js
javascript:VariableShadowing | CRITICAL | fiware-perseo:lib/appContext.js
javascript:VariableShadowing | CRITICAL | fiware-perseo:lib/db.js
javascript:VariableShadowing | CRITICAL | fiware-perseo:lib/db.js
javascript:VariableShadowing | CRITICAL | fiware-perseo:lib/db.js
javascript:VariableShadowing | CRITICAL | fiware-perseo:lib/db.js
javascript:VariableShadowing | CRITICAL | fiware-perseo:lib/db.js
javascript:ElseIfWithoutElse | CRITICAL | fiware-perseo:lib/models/actions.js
javascript:UnusedFunctionArgument | MAJOR | fiware-perseo:lib/models/actions.js
javascript:UnusedFunctionArgument | MAJOR | fiware-perseo:lib/models/actions.js
javascript:UnusedFunctionArgument | MAJOR | fiware-perseo:lib/models/actions.js
javascript:VariableShadowing | CRITICAL | fiware-perseo:lib/models/actions.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/actions.js
javascript:ElseIfWithoutElse | CRITICAL | fiware-perseo:lib/models/actionsStore.js
javascript:ElseIfWithoutElse | CRITICAL | fiware-perseo:lib/models/entitiesStore.js
javascript:VariableShadowing | CRITICAL | fiware-perseo:lib/models/executionsStore.js
javascript:VariableShadowing | CRITICAL | fiware-perseo:lib/models/executionsStore.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/metrics.js
javascript:UnusedFunctionArgument | MAJOR | fiware-perseo:lib/models/noSignal.js
javascript:UnusedFunctionArgument | MAJOR | fiware-perseo:lib/models/noSignal.js
javascript:UnusedFunctionArgument | MAJOR | fiware-perseo:lib/models/noSignal.js
javascript:UnusedFunctionArgument | MAJOR | fiware-perseo:lib/models/noSignal.js
javascript:VariableShadowing | CRITICAL | fiware-perseo:lib/models/notices.js
javascript:ElseIfWithoutElse | CRITICAL | fiware-perseo:lib/models/paths.js
javascript:ElseIfWithoutElse | CRITICAL | fiware-perseo:lib/models/postAction.js
javascript:ElseIfWithoutElse | CRITICAL | fiware-perseo:lib/models/postAction.js
javascript:ElseIfWithoutElse | CRITICAL | fiware-perseo:lib/models/rules.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/rulesStore.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/smppAction.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/smppAction.js
javascript:UnusedFunctionArgument | MAJOR | fiware-perseo:lib/models/twitterAction.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/updateAction.js
javascript:VariableShadowing | CRITICAL | fiware-perseo:lib/models/updateAction.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/visualRules.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/visualRules.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/visualRules.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/visualRules.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/visualRules.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/visualRules.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/visualRules.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/visualRules.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/visualRules.js
javascript:SwitchWithoutDefault | CRITICAL | fiware-perseo:lib/models/visualRules.js
javascript:ElseIfWithoutElse | CRITICAL | fiware-perseo:lib/models/visualRules.js
javascript:ElseIfWithoutElse | CRITICAL | fiware-perseo:lib/models/visualRules.js
javascript:SwitchWithoutDefault | CRITICAL | fiware-perseo:lib/models/visualRules.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/visualRules.js
javascript:S1226 | MINOR | fiware-perseo:lib/models/visualRules.js
javascript:S1226 | MINOR | fiware-perseo:lib/myutils.js
javascript:S1226 | MINOR | fiware-perseo:lib/myutils.js
javascript:S1226 | MINOR | fiware-perseo:lib/myutils.js
javascript:S1226 | MINOR | fiware-perseo:lib/myutils.js
javascript:ElseIfWithoutElse | CRITICAL | fiware-perseo:lib/myutils.js
javascript:S1226 | MINOR | fiware-perseo:lib/myutils.js
javascript:UnusedFunctionArgument | MAJOR | fiware-perseo:lib/myutils.js
javascript:S1226 | MINOR | fiware-perseo:lib/myutils.js
javascript:UnusedFunctionArgument | MAJOR | fiware-perseo:lib/perseo.js
javascript:VariableShadowing | CRITICAL | fiware-perseo:lib/perseo.js
javascript:S1135 | INFO | fiware-perseo:lib/routes/actionsController.js
javascript:VariableShadowing | CRITICAL | fiware-perseo:lib/routes/actionsController.js
javascript:S1226 | MINOR | fiware-perseo:lib/routes/visualRulesController.js
javascript:CommentedCode | MAJOR | fiware-perseo:lib/routes/visualRulesController.js

There are an error in Travis CI with Mongo connection, but all the issues detected by Travis have been corrected.

### CODE REVIEW

**Comment by fiqare-emergya-dev** [comment]
Swagger is provided for HTTP protocol in:
<server_host>:9090/api-docs

- Note: npm install is needed

File: .snyk [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Snyk (https://snyk.io) policy file, patches or ignores known vulnerabilities.
```
Comment (fgalan) [inline .snyk]: What is this file?

File: .travis.yml [hunk 2]
```diff
@@ -8,7 +8,7 @@ node_js:
 
 branches:
   only:
-    - master
+    - fiqare-perseo-fe-improvements
```
Comment (fgalan) [inline .travis.yml]: This shouldn't be changed in .travis.yml. Please, rollback the change.

File: README.md [hunk 3]
```diff
@@ -26,6 +26,7 @@ CEP implementation for IoT platform
 -   [Data model](documentation/models.md)
 -   [Available actions for PEP](documentation/pep_actions.md)
 -   [Documentation for developers of perseo](documentation/development.md)
+-   [Development documentation (Third Iteration)](documentation/eslint.md)
```
Comment (fgalan) [inline README.md]: A new section has been added in the ToC of the README.md file, but the actual section is not there...

Comment (fgalan) [reply README.md]: I have realized it points to another file. Thus, NTC. Sorry the noise.

File: config.js [hunk 4]
```diff
@@ -29,7 +29,7 @@ var config = {};
  * Default log level. Can be one of: 'DEBUG', 'INFO', 'WARN', 'ERROR', 'FATAL'
  * @type {string}
  */
-config.logLevel = 'INFO';
```
Comment (fgalan) [inline config.js]: Could you elaboratate on the changes in this file, please? Are they really needed?

File: documentation/eslint.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
+# Development documentation (Third Iteration)
```
Comment (fgalan) [inline documentation/eslint.md]: What does means "Third Iteration" here?

File: lib/alarm.js [hunk 6]
```diff
@@ -29,27 +29,29 @@ var util = require('util'),
 
 function raise(alarm, context, message) {
     var state = alarms[alarm];
-    context = (process.domain && process.domain.context) || {};
+    var contextTemp = (process.domain && process.domain.context) || {};
```
Comment (fgalan) [inline lib/alarm.js]: Could you elaborate on the usage of contextTemp and messageTemp instead of context and message here?

File: lib/db.js [hunk 7]
```diff
@@ -51,15 +51,10 @@ function getDbAux(url, component, callback) {
     client.connect(
         url,
         {
-            db: {
```
Comment (fgalan) [inline lib/db.js]: Is this releated with the new syntax used by mongodb module recent versions?

File: lib/models/entitiesStore.js [hunk 8]
```diff
@@ -52,6 +52,9 @@ function findSilentEntities(service, subservice, ruleData, func, callback) {
         } catch (e) {
             return callback(e, null);
         }
+    } else {
+    	//Added default else clause
```
Comment (fgalan) [inline lib/models/entitiesStore.js]: Under-indented

File: lib/models/notices.js [hunk 9]
```diff
@@ -249,7 +249,7 @@ function processCBv2Notice(service, subservice, ncr, ix) {
         var attrList = ncr.data[ix];
         for (var attr in attrList) {
             // Exclude id and type. NGSIv2
-            if (attr === 'id' || attr === 'type') {
+            if (attr !== 'id' && attr !== 'type') {
```
Comment (fgalan) [inline lib/models/notices.js]: Not fully sure both logical condition are equivalent... Or a bug in the logic has been solved with the change?

File: lib/models/postAction.js [hunk 10]
```diff
@@ -58,6 +62,10 @@ function doIt(action, event, callback) {
         } else if (options.text) {
             requestOptions.body = options.text;
         }
+        else {
+        	//Added default else clause
```
Comment (fgalan) [inline lib/models/postAction.js]: Over-indented line

File: lib/models/visualRules.js [hunk 11]
```diff
@@ -225,10 +237,15 @@ function vr2rule(cardRule) {
                             case 'timeElapsed': // Minimal interval for actions
                                 action.interval = card.timeData.interval;
                                 break;
+                            default:
+                            //nothing to do
```
Comment (fgalan) [inline lib/models/visualRules.js]: Under-idented line.

File: lib/perseo.js [hunk 12]
```diff
@@ -98,6 +100,22 @@ function start(callbackStart) {
         versionRoutes.AddTo(app);
         metricsRoutes.AddTo(app);
 
+        const options = {
```
Comment (fgalan) [inline lib/perseo.js]: Could you elaborate on the swagger-related changes, please?

Anyway, version should be a fixed value. It should be got from packages.json.

File: lib/routes/actionsController.js [hunk 13]
```diff
@@ -113,7 +113,6 @@ function PostAction(req, resp) {
     } else {
         delay = config.slaveDelay;
     }
-    // TODO: Tricky ... To refactor?
```
Comment (fgalan) [inline lib/routes/actionsController.js]: It seems to be a mark of pending technical debt. The refactor has not been done so I understand it should keep.

File: package.json [hunk 14]
```diff
@@ -21,14 +21,14 @@
   },
   "scripts": {
     "clean": "rm -rf package-lock.json && rm -rf node_modules && rm -rf coverage",
-    "test": "mocha --recursive 'test/**/*.js' --reporter spec --timeout 3000 --ui bdd --exit",
```
Comment (fgalan) [inline package.json]: Could you elaborate on this changes, please?

File: package.json [hunk 15]
```diff
@@ -53,7 +53,8 @@
     "textlint-rule-common-misspellings": "~1.0.1",
     "textlint-rule-no-dead-link": "~4.4.1",
     "textlint-rule-terminology": "~1.1.30",
-    "textlint-rule-write-good": "~1.6.2"
+    "textlint-rule-write-good": "~1.6.2",
+    "snyk": "^1.145.0"
```
Comment (fgalan) [inline package.json]: For stability reasons, we prefer to use ~ instead of ^. Please, use the former ones in the added dependecies.

File: test/utils/utilsT.js [hunk 16]
```diff
@@ -220,7 +220,7 @@ function addEntity(tenant, entity, callback) {
     );
 }
 function configTest() {
-    config.mongo.url = 'mongodb://localhost:27017/perseo_testing';
```
Comment (fgalan) [inline test/utils/utilsT.js]: Why this change?

**Comment by fgalan** [comment]
Thank you very much for your contribution! Apart from the in-line comment provided to the code changes in the PR, please take into account the following comments:

* This PR should use master as base branch
* Most of the changes are related with variable name changes. Could you elaborate on the need of these changes, please? We prefer not changing variable names if they are not providing any improvement (they make the PR reviewing process much more complex)
* It would be great if you could squash & rebase your PR to pass from 73 commits to 1 commit (or a small number)

**Comment by fgalan** [comment]
Continues in PR #419
